probe-rs-cli-util/src/common_options.rs

```diff
@@ -0,0 +1,453 @@
+        })?;
+
+        Ok(session)
+    }
+
+    /// Convenience method that attaches to the specified probe, target,
+    /// and target session.
+    pub fn create_session(&self) -> Result<Session, OperationError> {
+        let target = self.get_target_selector()?;
+        let probe = self.attach_probe()?;
+        let session = self.attach_session(probe, target)?;
+
+        Ok(session)
+    }
+
+    /// Builds a new flash loader for the given target and ELF. This
+    /// will check the ELF for validity and check what pages have to be
+    /// flashed etc.
+    pub fn build_flashloader(
+        &self,
+        target: &Option<TargetSelector>,
+        session: &mut Session,
+        elf_path: &Path,
+    ) -> Result<FlashLoader, OperationError> {
+        let target = match target {
+            Some(TargetSelector::Specified(target)) => target,
+            _ => session.target(),
+        };
+
+        // Create the flash loader
+        let mut loader = FlashLoader::new(target.memory_map.to_vec(), target.source().clone());
+
+        // Add data from the ELF.
+        let mut file = File::open(&elf_path).map_err(|error| OperationError::FailedToOpenElf {
+            source: error,
+            path: format!("{}", elf_path.display()),
+        })?;
+
+        // Try and load the ELF data.
+        loader
+            .load_elf_data(&mut file)
+            .map_err(OperationError::FailedToLoadElfData)?;
+
+        Ok(loader)
+    }
+}
+
+#[derive(StructOpt, Debug)]
+pub struct CargoOptions {
+    #[structopt(name = "binary", long = "bin", hidden = true)]
+    pub bin: Option<String>,
+    #[structopt(name = "example", long = "example", hidden = true)]
+    pub example: Option<String>,
+    #[structopt(name = "package", short = "p", long = "package", hidden = true)]
+    pub package: Option<String>,
+    #[structopt(name = "release", long = "release", hidden = true)]
+    pub release: bool,
+    #[structopt(name = "target", long = "target", hidden = true)]
+    pub target: Option<String>,
+    #[structopt(name = "PATH", long = "manifest-path", hidden = true)]
+    pub manifest_path: Option<String>,
+    #[structopt(long, hidden = true)]
+    pub no_default_features: bool,
+    #[structopt(long, hidden = true)]
+    pub all_features: bool,
+    #[structopt(long, hidden = true)]
+    pub features: Vec<String>,
+}
+
+impl CargoOptions {
+    pub fn help_message(bin: &str) -> String {
+        format!(
+            r#"
+CARGO BUILD OPTIONS:
+
+    The following options are forwarded to 'cargo build':
+
+        --bin
+        --example
+    -p, --package
+        --release
+        --target
+        --manifest-path
+        --no-default-features
+        --all-features
+        --features
+
+    For example, if you run the command '{} --release',
+    this means that 'cargo build --release' will be called.
+"#,
+            bin
+        )
+    }
+
+    pub fn to_cargo_arguments(&self) -> Vec<String> {
```

Comment:
In `cargo-flash` we don't translate the options back into strings, we just keep the original arguments. 
Do you have a different implementation for calling `cargo` in mind?

Reply:
`cargo-flash` builds with [`build_artifact`](https://github.com/probe-rs/probe-rs/blob/master/probe-rs-cli-util/src/lib.rs#L76). Before `to_cargo_arguments`, [`argument_handling`](https://github.com/probe-rs/probe-rs/blob/master/probe-rs-cli-util/src/argument_handling.rs) could be used to get a fitting `[String]`. But there a list of strings to exclude had to be maintained.

I have a slighly different `build_artifact` in my own project, but I need a `[String]` there also.

Reply:
I get your question now. From `CARGO_HELP_MESSAGE` I presumed that one cannot forward arbitrary commands to cargo, hence `CargoOptions` -> `[String]` instead of `std::env::args()` -> filter -> `[String]`.  I believe maintaining `CargoOptions` will be easier than a list of arguments that should be filtered out.

Reply:
If we maintain `CargoOptions`, we have to support all possible `cargo build` options, otherwise they cannot be used. We would also have to check for unused arguments to make sure no arguments are lost.

I think it would be easier to just remove the arguments which are used by our tools, than to maintain a list of all cargo options.

Reply:
Hmm, you're right. I'll revert back to the previous method.

Reply:
See latest commit.